From 3ecdc4e93ea27c4ade565c1b8d0a64aafcf91761 Mon Sep 17 00:00:00 2001
From: yogurt <yusong1117.u@gmail.com>
Date: Sat, 8 Nov 2025 00:55:11 +0800
Subject: [PATCH 3/3] feat: bring up st7789v panel

---
 drivers/staging/fbtft/fb_st7789v.c | 92 +++++++++---------------------
 1 file changed, 27 insertions(+), 65 deletions(-)

diff --git a/drivers/staging/fbtft/fb_st7789v.c b/drivers/staging/fbtft/fb_st7789v.c
index 3a280cc1892c..67028adfb7b6 100644
--- a/drivers/staging/fbtft/fb_st7789v.c
+++ b/drivers/staging/fbtft/fb_st7789v.c
@@ -82,66 +82,33 @@ enum st7789v_command {
  */
 static int init_display(struct fbtft_par *par)
 {
+	par->fbtftops.reset(par); // 这里一定要调用！
 	/* turn off sleep mode */
-	write_reg(par, MIPI_DCS_EXIT_SLEEP_MODE);
+	write_reg(par, 0x11); // sleep out
 	mdelay(120);
-
-	/* set pixel format to RGB-565 */
-	write_reg(par, MIPI_DCS_SET_PIXEL_FORMAT, MIPI_DCS_PIXEL_FMT_16BIT);
-	if (HSD20_IPS)
-		write_reg(par, PORCTRL, 0x05, 0x05, 0x00, 0x33, 0x33);
-
-	else
-		write_reg(par, PORCTRL, 0x08, 0x08, 0x00, 0x22, 0x22);
-
-	/*
-	 * VGH = 13.26V
-	 * VGL = -10.43V
-	 */
-	if (HSD20_IPS)
-		write_reg(par, GCTRL, 0x75);
-	else
-		write_reg(par, GCTRL, 0x35);
-
-	/*
-	 * VDV and VRH register values come from command write
-	 * (instead of NVM)
-	 */
-	write_reg(par, VDVVRHEN, 0x01, 0xFF);
-
-	/*
-	 * VAP =  4.1V + (VCOM + VCOM offset + 0.5 * VDV)
-	 * VAN = -4.1V + (VCOM + VCOM offset + 0.5 * VDV)
-	 */
-	if (HSD20_IPS)
-		write_reg(par, VRHS, 0x13);
-	else
-		write_reg(par, VRHS, 0x0B);
-
-	/* VDV = 0V */
-	write_reg(par, VDVS, 0x20);
-
-	/* VCOM = 0.9V */
-	if (HSD20_IPS)
-		write_reg(par, VCOMS, 0x22);
-	else
-		write_reg(par, VCOMS, 0x20);
-
-	/* VCOM offset = 0V */
-	write_reg(par, VCMOFSET, 0x20);
-
-	/*
-	 * AVDD = 6.8V
-	 * AVCL = -4.8V
-	 * VDS = 2.3V
-	 */
-	write_reg(par, PWCTRL1, 0xA4, 0xA1);
-
-	write_reg(par, MIPI_DCS_SET_DISPLAY_ON);
-
-	if (HSD20_IPS)
-		write_reg(par, MIPI_DCS_ENTER_INVERT_MODE);
-
+	write_reg(par, 0x36, 0x00);
+	write_reg(par, 0x3A, 0x05);
+	write_reg(par, 0xB2, 0x0C, 0x0C, 0x00, 0x33, 0x33);
+	write_reg(par, 0xB7, 0x56);
+	write_reg(par, 0xBB, 0x19);
+	write_reg(par, 0xC0, 0x2C);
+	write_reg(par, 0xC2, 0x01);
+	write_reg(par, 0xC3, 0x19);
+	write_reg(par, 0xC4, 0x20);
+	write_reg(par, 0xC6, 0x0F);
+	write_reg(par, 0xD0, 0xA7, 0xA1);
+	write_reg(par, 0xD0, 0xA4, 0xA1);
+	write_reg(par, 0xD6, 0xA1);
+	write_reg(par, 0xE0, 0xF0, 0x03, 0x08, 0x07, 0x09, 0x17, 0x2B, 0x33,
+		  0x41, 0x37, 0x12, 0x13, 0x2A, 0x2E);
+	write_reg(par, 0xE1, 0xF0, 0x05, 0x0A, 0x09, 0x07, 0x04, 0x2B, 0x33,
+		  0x40, 0x38, 0x14, 0x14, 0x28, 0x30);
+	write_reg(par, 0x21);
+	write_reg(par, 0x29); // display on
+	write_reg(par, 0x35); // enable TE
+	write_reg(par, 0x2A, 0x00, 0x00, 0x00, 0xEF); // column address 240
+	write_reg(par, 0x2B, 0x00, 0x00, 0x01, 0x3F); // raw address 320
+	write_reg(par, 0x2C); // memory write
 	return 0;
 }
 
@@ -244,9 +211,9 @@ static int set_gamma(struct fbtft_par *par, u32 *curves)
 static int blank(struct fbtft_par *par, bool on)
 {
 	if (on)
-		write_reg(par, MIPI_DCS_SET_DISPLAY_OFF);
+		write_reg(par, 0x10); // sleep in
 	else
-		write_reg(par, MIPI_DCS_SET_DISPLAY_ON);
+		write_reg(par, 0x11); // sleep out
 	return 0;
 }
 
@@ -254,13 +221,8 @@ static struct fbtft_display display = {
 	.regwidth = 8,
 	.width = 240,
 	.height = 320,
-	.gamma_num = 2,
-	.gamma_len = 14,
-	.gamma = HSD20_IPS_GAMMA,
 	.fbtftops = {
 		.init_display = init_display,
-		.set_var = set_var,
-		.set_gamma = set_gamma,
 		.blank = blank,
 	},
 };
-- 
2.34.1

